test/test_distributed.py

```diff
@@ -700,13 +734,79 @@ def test_all_gather_multigpu(self):
         self._test_all_gather_multigpu_helper(group, group_id, rank,
                                               rankToGPUMapping)
 
+    # END TO END TEST FOR DISTRIBUTEDDATAPARALLEL
+    def _test_DDP_helper(self, model, input_var, target, loss, optimizer):
+        model.train()
+        output = model(input_var)
+        l = loss(output, target)
+        l.backward()
+        optimizer.step()
+        return model
+
+    @unittest.skipIf(BACKEND != 'nccl' and BACKEND != 'gloo',
+                     "Only Nccl & Gloo backend support DistributedDataParallel")
+    @skip_if_no_cuda_distributed
+    @skip_if_no_multigpu
+    def test_DistributedDataParallel(self):
+        # Run a simple end to end DDP model, use result of single node model
+        # as baseline
+        group, group_id, rank = self._init_global_test()
+        rankToGPUMapping = self._init_multigpu_helper()
+
+        class Net(nn.Module):
+            def __init__(self):
+                super(Net, self).__init__()
+                self.features = nn.Linear(2, 4, bias=False)
+
+            def forward(self, x):
+                x = self.features(x)
+                return x
+
+        # cpu training setup
+        model = Net()
+        model_cpu = copy.deepcopy(model)
+        optimizer_cpu = torch.optim.SGD(model_cpu.parameters(), 0.1)
+
+        # DDP training setup
+        model_DDP = copy.deepcopy(model)
+        gpu_subset = list(rankToGPUMapping[rank])
+        model_DDP.cuda(gpu_subset[0])
+        model_DDP = nn.parallel.DistributedDataParallel(model_DDP,
+                                                        device_ids=gpu_subset)
+        optimizer_DDP = torch.optim.SGD(model_DDP.parameters(), 0.1)
+
+        # batch_size for DDP should be divisible by #GPU per node.
+        batch_size = len(gpu_subset) * int(WORLD_SIZE)
+        input_cpu = torch.randn(batch_size, 1, 2)
+        target = torch.randn(batch_size, 1, 4)
+        loss = nn.MSELoss()
+
+        model_cpu = self._test_DDP_helper(
+            model_cpu,
+            Variable(input_cpu, requires_grad=True),
+            Variable(target),
+            loss,
+            optimizer_cpu)
+
+        # DDP scatters subsets of input_cpu to GPUs
+        model_DDP = self._test_DDP_helper(
+            model_DDP,
+            Variable(input_cpu, requires_grad=True),
+            Variable(target.cuda(gpu_subset[0], non_blocking=True)),
+            loss,
+            optimizer_DDP)
+
+        self.assertEqual(model_cpu.features.weight,
+                         model_DDP.module.features.weight)
```

Comment:
To be honest I'm very surprised that they match. Every process will sample a random batch, and in the `model_cpu` case it will get some random gradient, while in the distributed case it will get the local gradient (same as for `model_cpu`) averaged with random gradients from other processes. Is it just me, or is the code wrong?